### PR TITLE
Fix/issue 115 editing and deleting conditions instantly syncs post experiment rule page when saved

### PR DESCRIPTION
**Pull request by ppratikcr7**
In this PR below things are done:

1. On editing a condition and saving it will not close the dialog box and let user to navigate to post experiment rule and select newly edited condition too.
2. On deleting a condition (while editing an experiment) which was assigned in post experiment rule will now default the selection to 'default' condition.
3. We have also kept a snackbar alert popup on save button on all modal pages while editing the experiment. This will let users know about changes being saved and not close the modal dialog box

### CODE REVIEW

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-design/experiment-design.component.ts [hunk 1]
```diff
@@ -185,6 +185,11 @@ export class ExperimentDesignComponent implements OnInit, OnChanges, OnDestroy {
 
   removeConditionOrPartition(type: string, groupIndex: number) {
     this[type].removeAt(groupIndex);
+    const deletedConditionCode = this.experimentInfo.conditions.find(condition => condition.order === groupIndex+1).id;
```
Comment (JD2455) [inline frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-design/experiment-design.component.ts:188]: This logic seems fragile like you are finding by relying on the `groupIndex + 1`. Can you update it in a better way than this?

Also there is a formatting issue. Please fix that as well

**Review by JD2455 — CHANGES_REQUESTED** [review]

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts [hunk 2]
```diff
@@ -22,26 +26,34 @@ export class ExperimentPostConditionComponent implements OnInit, OnChanges {
...
-          
-        } else {
-          isConditionExist = true;
-        }
-        this.experimentConditions = isConditionExist
-          ? this.experimentConditions
-          : [ ...this.experimentConditions, { value: value.conditionCode, id: value.id }];
-      });
-    }
+      this.experimentConditions = [
```
Comment (JD2455) [inline frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts:36]: Please fix the indentation issue.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts [hunk 3]
```diff
@@ -14,6 +17,7 @@ export class ExperimentPostConditionComponent implements OnInit, OnChanges {
   @Input() experimentInfo: ExperimentVM;
   @Input() newExperimentData: Partial<ExperimentVM>;
   @Output() emitExperimentDialogEvent = new EventEmitter<NewExperimentDialogData>();
+  @Input() experiment: ExperimentVM;
```
Comment (JD2455) [inline frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts:20]: What's the purpose of taking experiment having `ExperimentVM` as you already have experimentInfo with the same type which you are getting from the parent component?

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/new-experiment/new-experiment.component.ts [hunk 4]
```diff
@@ -57,18 +58,22 @@ export class NewExperimentComponent implements OnInit {
...
         };
+        this.openSnackBar();
         this.experimentService.updateExperiment(this.newExperimentData);
-        this.onNoClick();
         break;
     }
   }
 
+  openSnackBar() {
+    this._snackBar.open(`Experiment changes are updated successfully!`, null, { duration: 2000 })
```
Comment (JD2455) [inline frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/new-experiment/new-experiment.component.ts:75]: Can you please take this message from the en.json file?

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/new-experiment/new-experiment.component.ts [hunk 5]
```diff
@@ -57,18 +58,22 @@ export class NewExperimentComponent implements OnInit {
...
+        this.onNoClick();
       case NewExperimentDialogEvents.SAVE_DATA:
         this.newExperimentData = {
           ...this.experimentInfo,
           ...this.newExperimentData,
           ...formData
         };
+        this.openSnackBar();
         this.experimentService.updateExperiment(this.newExperimentData);
-        this.onNoClick();
```
Comment (JD2455) [inline frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/new-experiment/new-experiment.component.ts:67]: If you are removing `this.onNoClick` how will the new experiment or edit experiment dialog be closed after the saving data on the last tab? Is it something that the user have to click on the `CLOSE` from the every stepper?

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts [hunk 6]
```diff
@@ -22,26 +26,34 @@ export class ExperimentPostConditionComponent implements OnInit, OnChanges {
...
-        this.experimentConditions = isConditionExist
-          ? this.experimentConditions
-          : [ ...this.experimentConditions, { value: value.conditionCode, id: value.id }];
-      });
-    }
+      this.experimentConditions = [
+        { value: 'default', id: 'default' }
+      ];
+      if (this.experimentInfo) {
+        this.newExperimentStatSub = this.experimentService.selectExperimentById(this.experimentInfo.id).subscribe(stat => {
```
Comment (JD2455) [inline frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts:40]: Can you please update the variable name from the `this.newExperimentStatSub` to `this.experimentSub` as the selector is returning the experiment metadata.

Also the variable from `state` to experimentData` may be?

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/new-experiment/new-experiment.component.ts (hunk 5 above)
Comment (ppratikcr7) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/new-experiment/new-experiment.component.ts:67]: In the last tab it gets auto closed on pressing `create` or `update` button as it is handled in "UPDATE_EXPERIMENT" switch case. On others tab user needs to press `CLOSE` button.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts (hunk 6 above)
Comment (ppratikcr7) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts:40]: done

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts (hunk 2 above)
Comment (ppratikcr7) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts:36]: done

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts (hunk 3 above)
Comment (ppratikcr7) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-post-condition/experiment-post-condition.component.ts:20]: removed

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/new-experiment/new-experiment.component.ts (hunk 4 above)
Comment (ppratikcr7) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/new-experiment/new-experiment.component.ts:75]: done

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-design/experiment-design.component.ts (hunk 1 above)
Comment (ppratikcr7) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/experiment-design/experiment-design.component.ts]: GroupIndex starts with 0 and order starts from 1. Resolved minor edge case in latest commit.

**Review by JD2455 — APPROVED** [review]

**Review by VivekFitkariwala — APPROVED** [review]

**Review by msandbothe — APPROVED** [review]

**Pull request merged**